ethcore/src/block.rs

```diff
@@ -424,31 +418,9 @@ impl<'x> OpenBlock<'x> {
 
 	/// Turn this into a `LockedBlock`.
 	pub fn close_and_lock(self) -> Result<LockedBlock, Error> {
-		let mut s = self;
-
-		s.engine.on_close_block(&mut s.block)?;
-		s.block.state.commit()?;
-
-		if s.block.header.transactions_root().is_zero() || s.block.header.transactions_root() == &KECCAK_NULL_RLP {
-			s.block.header.set_transactions_root(ordered_trie_root(s.block.transactions.iter().map(|e| e.rlp_bytes())));
-		}
-		if s.block.header.uncles_hash().is_zero() || s.block.header.uncles_hash() == &KECCAK_EMPTY_LIST_RLP {
-			let uncle_bytes = encode_list(&s.block.uncles);
-			s.block.header.set_uncles_hash(keccak(&uncle_bytes));
-		}
-		if s.block.header.receipts_root().is_zero() || s.block.header.receipts_root() == &KECCAK_NULL_RLP {
-			s.block.header.set_receipts_root(ordered_trie_root(s.block.receipts.iter().map(|r| r.rlp_bytes())));
-		}
-
-		s.block.header.set_state_root(s.block.state.root().clone());
-		s.block.header.set_log_bloom(s.block.receipts.iter().fold(Bloom::zero(), |mut b, r| {
-			b.accrue_bloom(&r.log_bloom);
-			b
-		}));
-		s.block.header.set_gas_used(s.block.receipts.last().map_or_else(U256::zero, |r| r.gas_used));
-
+		let closed = self.close()?;
```

Comment:
Just want to note that this will have one additional `clone` of `unclosed_state` (https://github.com/paritytech/parity-ethereum/blob/a19faf01f347e7f485a1e1a82c47e7715e132dba/ethcore/src/block.rs#L397), which is immediately dropped. Not sure whether it's optimized away by compiler.

Reply:
Or maybe we can do it the other way around:

```
pub fn close(self) -> Result<ClosedBlock, Error> {
  let unclosed_state = self.block.state.clone();
  let locked = self.close_and_lock()?;
  
  Ok(ClosedBlock {
    block: locked.block,
    unclosed_state,
  })
}
```